File: ruby/triangle/triangle.rb
class Triangle
  attr_accessor :side_a, :side_b, :side_c
  def initialize(a, b, c)
    @side_a = a
    @side_b = b
    @side_c = c
  end

  def kind
    raise TriangleError if error?
    return :equilateral if equilateral?
    return :isosceles   if isosceles?
    return :scalene
  end

  private

  def error?
    triangle_inequality?(side_a, side_b, side_c) ||
    triangle_inequality?(side_b, side_c, side_a) ||
    triangle_inequality?(side_a, side_c, side_b)
  end

  def triangle_inequality?(x, y, z)
    x + y <= z
  end

  def equilateral?
    side_a == side_b && side_b === side_c
  end

  def isosceles?
    side_a == side_b || side_b == side_c || side_a == side_c
  end
end

class TriangleError < StandardError
end

File: python/triangle/triangle.py
class Triangle:

    EQUILATERAL = "equilateral"
    ISOSCELES = "isosceles"
    SCALENE = "scalene"

    def __init__(self, a, b, c):
        self.a = a
        self.b = b
        self.c = c
        if self.error():
            raise TriangleError

    def kind(self):
        if self.equilateral():
            return self.EQUILATERAL
        if self.isosceles():
            return self.ISOSCELES
        return self.SCALENE

    def equilateral(self):
        return self.a == self.b == self.c

    def isosceles(self):
        return self.a == self.b or self.b == self.c or self.a == self.c

    def error(self):
        return self.negative_sides() or self.triangle_inequality()

    def negative_sides(self):
        return self.a <= 0 or self.b <= 0 or self.c <= 0

    def triangle_inequality(self):
        return (self.a + self.b <= self.c or
                self.b + self.c <= self.a or
                self.a + self.c <= self.b)


class TriangleError(Exception):
    pass
